File: modules/project/src/main/java/com/falsepattern/zigbrains/project/toolchain/ZigToolchainEnvironmentSerializable.java
/*
 * Copyright 2023 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.toolchain;

import com.google.gson.annotations.SerializedName;

public record ZigToolchainEnvironmentSerializable(
        @SerializedName("zig_exe")
        String zigExecutable,
        @SerializedName("lib_dir")
        String libDirectory,
        @SerializedName("std_dir")
        String stdDirectory,
        @SerializedName("global_cache_dir")
        String globalCacheDirectory,
        @SerializedName("version")
        String version,
        @SerializedName("target")
        String target
) {}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/project/toolchain/ZigToolchainEnvironmentSerializable.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2024 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */
package com.falsepattern.zigbrains.project.toolchain

import com.intellij.openapi.util.io.toNioPathOrNull
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.nio.file.Path


@JvmRecord
@Serializable
data class ZigToolchainEnvironmentSerializable(
    @SerialName("zig_exe") val zigExecutable: String,
    @SerialName("std_dir") val stdDirectory: String,
    @SerialName("global_cache_dir") val globalCacheDirectory: String,
    @SerialName("lib_dir") val libDirectory: String,
    @SerialName("version") val version: String,
    @SerialName("target") val target: String
) {
    fun stdPath(toolchain: LocalZigToolchain): Path? {
        val path = stdDirectory.toNioPathOrNull() ?: return null
        if (path.isAbsolute)
            return path

        val resolvedPath = toolchain.location.resolve(path)
        if (resolvedPath.isAbsolute)
            return resolvedPath

        return null
    }
}
